### PR TITLE
Add `character/char` data type and fix alias for `byte`

### LINKED ISSUE #12512: Ability to cast `TEXT` values to `CHAR`
**Use case**:
PostgreSQL compatibility / Schema retrieval on JDBC/ODBC (Progress) driver site

**Feature description**:

```sql
Select 'A'::CHAR 
-- returns A

Select 'AB'::CHAR    
-- returns A 
```

### PR DESCRIPTION
**Pull request by seut**
## Summary of the changes / Why this improves CrateDB

The single-byte `byte` data type was falsely alias with `char` while it should have aliased with `"char"` (Note the double quotes).
The first commit fixes this alias, while the second commit implements a dedicated `CharacterType`.

See also https://www.postgresql.org/docs/14/datatype-character.html and specially https://www.postgresql.org/docs/14/datatype-character.html#DATATYPE-CHARACTER-SPECIAL-TABLE.

Fixes #12512.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by seut** [comment]
I've removed review request for now, since 1 test fails now (due some recent changes in my branch). I'll look into this asap.

**Comment by seut** [comment]
retest this please, unrelated kill test failures

File: docs/appendices/release-notes/unreleased.rst [hunk 1]
```diff
@@ -71,6 +71,10 @@ Breaking Changes
   ``pg_index`` tables now return ``TEXT`` instead of ``ARRAY(OBJECT)`` for
   improved compatibility with PostgreSQL.
 
+- The single byte ``byte`` data type was falsely exposed as the ``char``
+  data type. It will now be correctly exposed as the special single byte
+  `"char"`.
```
Comment (mfussenegger) [inline docs/appendices/release-notes/unreleased.rst:76]:
```suggestion
  ``"char"``.
```

File: server/src/main/java/io/crate/protocols/postgres/types/CharacterType.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+package io.crate.protocols.postgres.types;
+
+import java.nio.charset.StandardCharsets;
+
+import javax.annotation.Nonnull;
+
+import io.crate.types.DataTypes;
+import io.netty.buffer.ByteBuf;
+
+class CharacterType extends PGType<Object> {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/protocols/postgres/types/CharacterType.java:31]: Could this be changed to extend `PGType<String>` ?

File: server/src/main/java/io/crate/types/CharacterType.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    private final int lengthLimit;
+
+    public CharacterType(StreamInput in) throws IOException {
+        if (in.getVersion().onOrAfter(Version.V_5_0_0)) {
+            lengthLimit = in.readInt();
+        } else {
+            lengthLimit = 1;
+        }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/CharacterType.java:72]: The type is new in 5.0.0, why do we need version checks here?

File: server/src/main/java/io/crate/types/CharacterType.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    private String pad(String value) {
+        char[] buffer = new char[lengthLimit];
+        System.arraycopy(value.toCharArray(), 0, buffer, 0, value.length());
+        for (int i = value.length(); i < lengthLimit; i++) {
+            buffer[i] = ' ';
+        }
+        return String.valueOf(buffer);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/CharacterType.java:134]: There would also be a `padEnd` in `StringUtils`.  (and I think a regular string concat might even be faster, but didn't verify)

File: server/src/main/java/io/crate/types/CharacterType.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        if (string.length() <= lengthLimit()) {
+            return string;
+        } else {
+            return string.substring(0, lengthLimit());
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        if (out.getVersion().onOrAfter(Version.V_5_0_0)) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/CharacterType.java:161]: Probably not needed, see other comment.

File: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java [hunk 6]
```diff
@@ -193,6 +203,8 @@ public String typeName() {
     private int ignoreAbove;
     private Integer lengthLimit;
     private String nullValue;
+    @Nullable
+    private Boolean blankPadding;
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:207]: Could we default this to `false` and make it non-nullable + primitive? Would reduce the number of possible states.

**Review by mfussenegger — COMMENTED** [review]
Cool stuff. I left a couple of minor commits - overall looks very good

**Comment by mfussenegger** [comment]
Oh I think we also need to extend the list here: https://github.com/crate/crate/blob/aefd4a221d835f35a917cdbf4bc2249daf010a0a/docs/interfaces/http.rst#L239-L288

**Comment by seut** [comment]
> Oh I think we also need to extend the list here: 

ah thanks, good catch!

File: server/src/main/java/io/crate/types/CharacterType.java (hunk 4 above)
Comment (seut) [reply server/src/main/java/io/crate/types/CharacterType.java:134]: Wasn't aware of this. I've run a quick benchmark, comparing all 3 variants with a string `foo` (3 chars) and padding to 10 chars with whitespaces:
```
Benchmark                           Mode  Cnt         Score         Error  Units
CharacterTypeTest.padConcatRepeat  thrpt   25  95202620,469 ± 3391085,804  ops/s
CharacterTypeTest.padCustom        thrpt   25  46869678,245 ±  933217,198  ops/s
CharacterTypeTest.padStringUtils   thrpt   25  35043663,237 ±  835865,710  ops/s
```

Whereas `padConcatRepeat` is implemented like this:
```"foo" + " ".repeat(7);```

So seems that concat + repeat wins, even a bit surprising as it uses a similar approach than my custom pad function.

Comment (mfussenegger) [reply server/src/main/java/io/crate/types/CharacterType.java:134]: > Based on these results, shall we use concat + repeat inside StringUtils.padEnd as well?

I'd say yes.

> even a bit surprising as it uses a similar approach than my custom pad function.

I think the JDK added some intrinsics for the JIT/c2 compiler at some point, but I could be mixing things up.

**Review by mfussenegger — APPROVED** [review]

**Pull request merged**